src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseImportFromFileService.java

```diff
@@ -0,0 +1,193 @@
+     * It first extracts the contents of the zip file, then creates a programming exercise (same process as creating a new one),
+     * then deletes the template content initially pushed to the repositories and copies over the extracted content
+     *
+     * @param programmingExerciseForImport the programming exercise that should be imported
+     * @param zipFile                      the zip file that contains the exercise
+     * @param course                       the course to which the exercise should be added
+     * @return the imported programming exercise
+     **/
+    public ProgrammingExercise importProgrammingExerciseFromFile(ProgrammingExercise programmingExerciseForImport, MultipartFile zipFile, Course course)
+            throws IOException, GitAPIException, URISyntaxException {
+        Path importExerciseDir = Files.createTempDirectory("imported-exercise-dir");
+        Path exerciseFilePath = Files.createTempFile(importExerciseDir, "exercise-for-import", ".zip");
+        if (!"zip".equals(FileNameUtils.getExtension(zipFile.getOriginalFilename()))) {
+            throw new BadRequestAlertException("The file is not a zip file", "programmingExercise", "fileNotZip");
+        }
+        zipFile.transferTo(exerciseFilePath);
+        zipFileService.extractZipFileRecursively(exerciseFilePath);
+        checkRepositoriesExist(importExerciseDir);
+        var oldShortName = getProgrammingExerciseFromDetailsFile(importExerciseDir).getShortName();
+        programmingExerciseService.validateNewProgrammingExerciseSettings(programmingExerciseForImport, course);
+        ProgrammingExercise importedProgrammingExercise = programmingExerciseService.createProgrammingExercise(programmingExerciseForImport);
+        if (Boolean.TRUE.equals(programmingExerciseForImport.isStaticCodeAnalysisEnabled())) {
+            staticCodeAnalysisService.createDefaultCategories(importedProgrammingExercise);
+        }
+        importRepositoriesFromFile(importedProgrammingExercise, importExerciseDir, oldShortName);
+        return importedProgrammingExercise;
+    }
+
+    private void importRepositoriesFromFile(ProgrammingExercise newExercise, Path basePath, String oldExerciseShortName) throws IOException, GitAPIException, URISyntaxException {
+        Repository templateRepo = gitService.getOrCheckoutRepository(new VcsRepositoryUrl(newExercise.getTemplateRepositoryUrl()), false);
+        Repository solutionRepo = gitService.getOrCheckoutRepository(new VcsRepositoryUrl(newExercise.getSolutionRepositoryUrl()), false);
+        Repository testRepo = gitService.getOrCheckoutRepository(new VcsRepositoryUrl(newExercise.getTestRepositoryUrl()), false);
+
+        copyImportedExerciseContentToRepositories(templateRepo, solutionRepo, testRepo, basePath);
+        replaceImportedExerciseShortName(Map.of(oldExerciseShortName, newExercise.getShortName()), List.of("gradle-wrapper.jar"), templateRepo, solutionRepo, testRepo);
```

Comment:
```suggestion
        replaceImportedExerciseShortName(Map.of(oldExerciseShortName, newExercise.getShortName()), templateRepo, solutionRepo, testRepo);
```
Instead of passing the list here as parameter I’d define a `private static final List<String> SHORT_NAME_REPLACEMENT_EXCLUSIONS = List.of("gradle-wrapper.jar")`. It is likely that more such exception might occur in the future.